// salesforce-alm/dist/lib/community/connect/CommunityCreateResource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@salesforce/core");
core_1.Messages.importMessagesDirectory(__dirname);
const communityMessages = core_1.Messages.loadMessages('salesforce-alm', 'community_commands');
const MESSAGE_KEY = 'message';
const NAME_KEY = 'name';
const ACTION_KEY = 'action';
/**
 * A connect api resource for creating a community
 */
class CommunityCreateResource {
    constructor(flags, ux) {
        this.flags = flags;
        this.ux = ux;
    }
    handleSuccess(result) {
        const response = {
            message: communityMessages.getMessage('create.response.createMessage'),
            name: result[NAME_KEY],
            action: communityMessages.getMessage('create.response.action')
        };
        const columns = [NAME_KEY, MESSAGE_KEY, ACTION_KEY];
        this.ux.styledHeader(communityMessages.getMessage('create.response.styleHeader'));
        this.ux.table([response], columns);
        return response;
    }
    handleError(error) {
        throw error;
    }
    async fetchRelativeConnectUrl() {
        return '/connect/communities';
    }
    getRequestMethod() {
        return 'POST';
    }
    async fetchPostParams() {
        const params = {
            name: this.flags.name,
            urlPathPrefix: this.flags.urlpathprefix,
            templateName: this.flags.templatename,
            description: this.flags.description
        };
        return JSON.stringify(params);
    }
}
exports.CommunityCreateResource = CommunityCreateResource;
