// salesforce-alm/dist/lib/community/connect/CommunityTemplatesResource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@salesforce/core");
core_1.Messages.importMessagesDirectory(__dirname);
const community = core_1.Messages.loadMessages('salesforce-alm', 'community_commands');
/**
 * A connect api resource for fetching community templates available to context user
 */
class CommunityTemplatesResource {
    constructor(ux) {
        this.ux = ux;
    }
    async fetchRelativeConnectUrl() {
        return '/connect/communities/templates';
    }
    async fetchPostParams() {
        return JSON.stringify({});
    }
    getRequestMethod() {
        return 'GET';
    }
    handleSuccess(result) {
        const columns = ['templateName', 'publisher'];
        this.ux.styledHeader(community.getMessage('list.response.styledHeader'));
        this.ux.table(result.templates, columns);
        this.ux.log();
        this.ux.log(community.getMessage('list.response.TotalField'), result.total.toString());
        return result;
    }
    handleError(error) {
        throw error;
    }
}
exports.CommunityTemplatesResource = CommunityTemplatesResource;
