// @lwc/engine/lib/framework/main.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
// Polyfills
require("../polyfills/proxy-concat/main");
require("../polyfills/aria-properties/main");
// TODO: #1296 - Revisit these exports and figure out a better separation
var upgrade_1 = require("./upgrade");
exports.createElement = upgrade_1.createElement;
var def_1 = require("./def");
exports.getComponentDef = def_1.getComponentDef;
exports.isComponentConstructor = def_1.isComponentConstructor;
exports.getComponentConstructor = def_1.getComponentConstructor;
var base_lightning_element_1 = require("./base-lightning-element");
exports.LightningElement = base_lightning_element_1.BaseLightningElement;
var services_1 = require("./services");
exports.register = services_1.register;
var membrane_1 = require("./membrane");
exports.unwrap = membrane_1.unwrap;
var secure_template_1 = require("./secure-template");
exports.registerTemplate = secure_template_1.registerTemplate;
exports.sanitizeAttribute = secure_template_1.sanitizeAttribute;
var component_1 = require("./component");
exports.registerComponent = component_1.registerComponent;
var register_1 = require("./decorators/register");
exports.registerDecorators = register_1.registerDecorators;
var vm_1 = require("./vm");
exports.isNodeFromTemplate = vm_1.isNodeFromTemplate;
var api_1 = require("./decorators/api");
exports.api = api_1.default;
var track_1 = require("./decorators/track");
exports.track = track_1.default;
var readonly_1 = require("./decorators/readonly");
exports.readonly = readonly_1.default;
var wire_1 = require("./decorators/wire");
exports.wire = wire_1.default;
var decorate_1 = require("./decorators/decorate");
exports.decorate = decorate_1.default;
var wc_1 = require("./wc");
exports.buildCustomElementConstructor = wc_1.buildCustomElementConstructor;
