// @lwc/synthetic-shadow/lib/env/dom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const language_1 = require("../shared/language");
const ShadowRootHostGetter = typeof window.ShadowRoot !== 'undefined'
    ? language_1.getOwnPropertyDescriptor(window.ShadowRoot.prototype, 'host').get
    : () => {
        throw new Error('Internal Error: Missing ShadowRoot');
    };
exports.ShadowRootHostGetter = ShadowRootHostGetter;
const dispatchEvent = 'EventTarget' in window ? EventTarget.prototype.dispatchEvent : Node.prototype.dispatchEvent; // IE11
exports.dispatchEvent = dispatchEvent;
const isNativeShadowRootAvailable = typeof window.ShadowRoot !== 'undefined';
exports.isNativeShadowRootAvailable = isNativeShadowRootAvailable;
const eventTargetGetter = language_1.getOwnPropertyDescriptor(Event.prototype, 'target').get;
exports.eventTargetGetter = eventTargetGetter;
const eventCurrentTargetGetter = language_1.getOwnPropertyDescriptor(Event.prototype, 'currentTarget').get;
exports.eventCurrentTargetGetter = eventCurrentTargetGetter;
const focusEventRelatedTargetGetter = language_1.getOwnPropertyDescriptor(FocusEvent.prototype, 'relatedTarget').get;
exports.focusEventRelatedTargetGetter = focusEventRelatedTargetGetter;
