// salesforce-alm/dist/lib/org/sandbox/sandboxConstants.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const kit_1 = require("@salesforce/kit");
exports.SANDBOXDEF_SRC_SANDBOXNAME = 'SourceSandboxName';
class SandboxConstants {
}
exports.SandboxConstants = SandboxConstants;
SandboxConstants.DEFAULT_MAX_RETRIES = 0;
SandboxConstants.DEFAULT_POLL_INTERVAL = kit_1.Duration.seconds(30);
SandboxConstants.SANDBOX_INCOMPLETE_EXCEPTION_MESSAGE = 'INVALID_STATUS';
class SandboxEventNames {
}
exports.SandboxEventNames = SandboxEventNames;
SandboxEventNames.EVENT_STATUS = 'status';
SandboxEventNames.EVENT_ASYNCRESULT = 'asyncResult';
SandboxEventNames.EVENT_RESULT = 'result';
SandboxEventNames.EVENT_CLIENTID_NOTSUPPORT = 'clientIdNotSupported';
var SandboxStatus;
(function (SandboxStatus) {
    SandboxStatus["Pending"] = "Pending";
    SandboxStatus["PendingRemote"] = "Pending Remote Creation";
    SandboxStatus["Processing"] = "Processing";
    SandboxStatus["Completed"] = "Completed";
    SandboxStatus["Activating"] = "Activating";
})(SandboxStatus = exports.SandboxStatus || (exports.SandboxStatus = {}));
var SandboxType;
(function (SandboxType) {
    SandboxType["DEVELOPER"] = "DEVELOPER";
    SandboxType["FULL"] = "FULL";
    SandboxType["LDPS"] = "ADVANCED_DEV";
})(SandboxType = exports.SandboxType || (exports.SandboxType = {}));
