// salesforce-alm/dist/commands/force/apex/log/tail.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../../lib/force-cli/force-cli-messages");
const ToolbeltCommand_1 = require("../../../../ToolbeltCommand");
class ApexLogTailCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const { ApexLogTailCommand } = require('../../../../lib/apex/apexLogTailCommand');
        return await this.execLegacyCommand(new ApexLogTailCommand(), context);
    }
}
exports.ApexLogTailCommand = ApexLogTailCommand;
ApexLogTailCommand.theDescription = Messages.get('ApexLogTailDescription');
ApexLogTailCommand.longDescription = Messages.get('ApexLogTailLongDescription');
ApexLogTailCommand.help = Messages.get('ApexLogTailHelp');
ApexLogTailCommand.requiresProject = false;
ApexLogTailCommand.flagsConfig = {
    color: command_1.flags.boolean({
        char: 'c',
        required: false,
        hidden: false,
        description: Messages.get('ApexLogTailColorizeDescription'),
        longDescription: Messages.get('ApexLogTailColorizeLongDescription')
    }),
    debuglevel: command_1.flags.string({
        char: 'd',
        required: false,
        hidden: false,
        description: Messages.get('ApexLogTailDebugLevelDescription'),
        longDescription: Messages.get('ApexLogTailDebugLevelLongDescription')
    }),
    skiptraceflag: command_1.flags.boolean({
        char: 's',
        required: false,
        hidden: false,
        description: Messages.get('ApexLogTailSkipTraceFlagDescription'),
        longDescription: Messages.get('ApexLogTailSkipTraceFlagLongDescription')
    })
};
ApexLogTailCommand.requiresUsername = true;
