// salesforce-alm/dist/commands/force/org/snapshot/create.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../../lib/messages");
const ToolbeltCommand_1 = require("../../../../ToolbeltCommand");
const messages = Messages();
const consts = require("../../../../lib/core/constants");
class OrgSnapshotCreateCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const OrgSnapshotCreateCommand = require('../../../../lib/org/snapshot/orgSnapshotCreateCommand'); // eslint-disable-line global-require
        return this.execLegacyCommand(new OrgSnapshotCreateCommand(), context);
    }
}
exports.OrgSnapshotCreateCommand = OrgSnapshotCreateCommand;
OrgSnapshotCreateCommand.theDescription = messages.getMessage('createSnapshotCmdDescription', [], 'orgSnapshot');
OrgSnapshotCreateCommand.longDescription = messages.getMessage('createSnapshotCmdDescriptionLong', [], 'orgSnapshot');
OrgSnapshotCreateCommand.help = messages.getMessage('createSnapshotCmdHelp', [], 'orgSnapshot');
OrgSnapshotCreateCommand.showProgress = true;
OrgSnapshotCreateCommand.requiresProject = false;
OrgSnapshotCreateCommand.requiresDevhubUsername = true;
OrgSnapshotCreateCommand.orgType = consts.DEFAULT_DEV_HUB_USERNAME;
OrgSnapshotCreateCommand.flagsConfig = {
    sourceorg: command_1.flags.string({
        char: 'o',
        description: messages.getMessage('createSnapshotCmdSourceOrgDescription', [], 'orgSnapshot'),
        longDescription: messages.getMessage('createSnapshotCmdSourceOrgDescriptionLong', [], 'orgSnapshot'),
        required: true
    }),
    snapshotname: command_1.flags.string({
        char: 'n',
        description: messages.getMessage('createSnapshotCmdNameDescription', [], 'orgSnapshot'),
        longDescription: messages.getMessage('createSnapshotCmdNameDescriptionLong', [], 'orgSnapshot'),
        required: true
    }),
    description: command_1.flags.string({
        char: 'd',
        description: messages.getMessage('createSnapshotCmdDescriptionDescription', [], 'orgSnapshot'),
        longDescription: messages.getMessage('createSnapshotCmdDescriptionDescriptionLong', [], 'orgSnapshot'),
        required: false
    })
};
