// salesforce-alm/dist/commands/force/org/snapshot/delete.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../../lib/messages");
const ToolbeltCommand_1 = require("../../../../ToolbeltCommand");
const messages = Messages();
const consts = require("../../../../lib/core/constants");
class OrgSnapshotDeleteCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const OrgSnapshotDeleteCommand = require('../../../../lib/org/snapshot/orgSnapshotDeleteCommand'); // eslint-disable-line global-require
        return this.execLegacyCommand(new OrgSnapshotDeleteCommand(), context);
    }
}
exports.OrgSnapshotDeleteCommand = OrgSnapshotDeleteCommand;
OrgSnapshotDeleteCommand.theDescription = messages.getMessage('deleteSnapshotCmdDescription', [], 'orgSnapshot');
OrgSnapshotDeleteCommand.longDescription = messages.getMessage('deleteSnapshotCmdDescriptionLong', [], 'orgSnapshot');
OrgSnapshotDeleteCommand.help = messages.getMessage('deleteSnapshotCmdHelp', [], 'orgSnapshot');
OrgSnapshotDeleteCommand.showProgress = true;
OrgSnapshotDeleteCommand.requiresProject = false;
OrgSnapshotDeleteCommand.requiresDevhubUsername = true;
OrgSnapshotDeleteCommand.orgType = consts.DEFAULT_DEV_HUB_USERNAME;
OrgSnapshotDeleteCommand.flagsConfig = {
    snapshot: command_1.flags.string({
        char: 's',
        description: messages.getMessage('deleteSnapshotCmdSnapshotDescription', [], 'orgSnapshot'),
        longDescription: messages.getMessage('deleteSnapshotCmdSnapshotDescriptionLong', [], 'orgSnapshot'),
        required: true
    })
};
