// salesforce-alm/dist/commands/force/org/snapshot/get.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../../lib/messages");
const ToolbeltCommand_1 = require("../../../../ToolbeltCommand");
const messages = Messages();
const consts = require("../../../../lib/core/constants");
class OrgSnapshotGetCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const OrgSnapshotGetCommand = require('../../../../lib/org/snapshot/orgSnapshotGetCommand'); // eslint-disable-line global-require
        return this.execLegacyCommand(new OrgSnapshotGetCommand(), context);
    }
}
exports.OrgSnapshotGetCommand = OrgSnapshotGetCommand;
OrgSnapshotGetCommand.theDescription = messages.getMessage('getSnapshotCmdDescription', [], 'orgSnapshot');
OrgSnapshotGetCommand.longDescription = messages.getMessage('getSnapshotCmdDescriptionLong', [], 'orgSnapshot');
OrgSnapshotGetCommand.help = messages.getMessage('getSnapshotCmdHelp', [], 'orgSnapshot');
OrgSnapshotGetCommand.showProgress = true;
OrgSnapshotGetCommand.requiresProject = false;
OrgSnapshotGetCommand.requiresDevhubUsername = true;
OrgSnapshotGetCommand.orgType = consts.DEFAULT_DEV_HUB_USERNAME;
OrgSnapshotGetCommand.flagsConfig = {
    snapshot: command_1.flags.string({
        char: 's',
        description: messages.getMessage('getSnapshotCmdSnapshotDescription', [], 'orgSnapshot'),
        longDescription: messages.getMessage('getSnapshotCmdSnapshotDescriptionLong', [], 'orgSnapshot'),
        required: true
    })
};
