// salesforce-alm/dist/commands/force/community/create.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@salesforce/command");
const ConnectExecutor_1 = require("../../../lib/connect/services/ConnectExecutor");
const CommunityCreateResource_1 = require("../../../lib/community/connect/CommunityCreateResource");
const core_1 = require("@salesforce/core");
core_1.Messages.importMessagesDirectory(__dirname);
const communityMessages = core_1.Messages.loadMessages('salesforce-alm', 'community_commands');
/**
 * A command to create a community.
 * This is just an sfdx wrapper around the community create connect endpoint
 */
class CommunityCreateCommand extends command_1.SfdxCommand {
    async run() {
        const createCommand = new CommunityCreateResource_1.CommunityCreateResource(this.flags, this.ux);
        return new ConnectExecutor_1.ConnectExecutor(createCommand, this.org).callConnectApi();
    }
}
exports.CommunityCreateCommand = CommunityCreateCommand;
CommunityCreateCommand.requiresUsername = true;
CommunityCreateCommand.help = communityMessages.getMessage('create.help');
CommunityCreateCommand.longDescription = communityMessages.getMessage('create.longDescription');
CommunityCreateCommand.description = communityMessages.getMessage('create.description') + '\n\n' + CommunityCreateCommand.help;
CommunityCreateCommand.flagsConfig = {
    name: command_1.flags.string({
        char: 'n',
        description: communityMessages.getMessage('create.flags.name.description'),
        longDescription: communityMessages.getMessage('create.flags.name.longDescription'),
        required: true
    }),
    templatename: command_1.flags.string({
        char: 't',
        description: communityMessages.getMessage('create.flags.templateName.description'),
        longDescription: communityMessages.getMessage('create.flags.templateName.longDescription'),
        required: true
    }),
    urlpathprefix: command_1.flags.string({
        char: 'p',
        description: communityMessages.getMessage('create.flags.urlPathPrefix.description'),
        longDescription: communityMessages.getMessage('create.flags.urlPathPrefix.longDescription'),
        required: true
    }),
    description: command_1.flags.string({
        char: 'd',
        description: communityMessages.getMessage('create.flags.description.description'),
        longDescription: communityMessages.getMessage('create.flags.description.longDescription'),
        required: false
    })
};
