// salesforce-alm/messages/mdapi_cancel.js
module.exports = {
  description: 'cancel a metadata deployment',
  longDescription: 'Cancels an asynchronous metadata deployment.',
  help: `Use this command to cancel a specified asynchronous metadata deployment. You can also specify a wait time (in minutes) to check for updates to the canceled deploy status.

Examples:

Deploy a directory of files to the org
  $ sfdx force:mdapi:deploy -d <directory>
Now cancel this deployment and wait two minutes
  $ sfdx force:mdapi:deploy:cancel -w 2

If you have multiple deployments in progress and want to cancel a specific one, specify the job ID
  $ sfdx force:mdapi:deploy:cancel -i <jobid>
Check the status of the cancel job
  $ sfdx force:mdapi:deploy:report`,

  jobId: 'job ID of the deployment you want to cancel; defaults to your most recent CLI deployment if not specified',
  jobIdLong: 'The job ID (requestId) of the deployment you want to cancel. If not specified, the default value is the ID of the most recent metadata deployment you ran using Salesforce CLI.',

  MissingRequiredParameter: 'A required parameter [jobid] is missing.',
  CancelFailed: 'The cancel command failed due to: %s'
};
