// salesforce-alm/dist/commands/force/package/hammertest/list.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const ToolbeltCommand_1 = require("../../../../ToolbeltCommand");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('salesforce-alm', 'package_hammertest_list');
class PackageHammertestListCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const HammerTestListCommand = require('../../../../lib/package/hammerTestListCommand');
        return this.execLegacyCommand(new HammerTestListCommand(), context);
    }
}
exports.PackageHammertestListCommand = PackageHammertestListCommand;
PackageHammertestListCommand.theDescription = messages.getMessage('cliDescription');
PackageHammertestListCommand.longDescription = messages.getMessage('cliLongDescription');
PackageHammertestListCommand.help = messages.getMessage('help');
PackageHammertestListCommand.requiresUsername = true;
PackageHammertestListCommand.requiresProject = true;
PackageHammertestListCommand.flagsConfig = {
    packageversionid: command_1.flags.id({
        char: 'i',
        description: messages.getMessage('packageversionid'),
        longDescription: messages.getMessage('packageversionidLong'),
        required: false
    })
};
