// salesforce-alm/dist/commands/force/auth/sfdxurl/store.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../../lib/messages");
const ToolbeltCommand_1 = require("../../../../ToolbeltCommand");
const messages = Messages();
const consts = require("../../../../lib/core/constants");
class AuthSfdxurlStoreCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const AuthCommand = require('../../../../lib/auth/authCommand');
        const authCommand = new AuthCommand();
        return this.execLegacyCommand(authCommand, context);
    }
}
exports.AuthSfdxurlStoreCommand = AuthSfdxurlStoreCommand;
AuthSfdxurlStoreCommand.theDescription = messages.getMessage('description', [], 'auth_sfdxurl');
AuthSfdxurlStoreCommand.longDescription = messages.getMessage('longDescription', [], 'auth_sfdxurl');
AuthSfdxurlStoreCommand.help = messages.getMessage('help', [consts.AUTH_URL_FORMAT1, consts.AUTH_URL_FORMAT2], 'auth_sfdxurl');
AuthSfdxurlStoreCommand.requiresProject = false;
AuthSfdxurlStoreCommand.flagsConfig = {
    sfdxurlfile: command_1.flags.filepath({
        char: 'f',
        description: messages.getMessage('file', [], 'auth_sfdxurl'),
        longDescription: messages.getMessage('fileLong', [], 'auth_sfdxurl'),
        required: true
    }),
    setdefaultdevhubusername: command_1.flags.boolean({
        char: 'd',
        description: messages.getMessage('setDefaultDevHub', [], 'auth'),
        longDescription: messages.getMessage('setDefaultDevHubLong', [], 'auth')
    }),
    setdefaultusername: command_1.flags.boolean({
        char: 's',
        description: messages.getMessage('setDefaultUsername', [], 'auth'),
        longDescription: messages.getMessage('setDefaultUsernameLong', [], 'auth')
    }),
    setalias: command_1.flags.string({
        char: 'a',
        description: messages.getMessage('setAlias', [], 'auth'),
        longDescription: messages.getMessage('setAliasLong', [], 'auth')
    }),
    noprompt: command_1.flags.boolean({
        char: 'p',
        description: messages.getMessage('noPrompt', [], 'demoMode'),
        longDescription: messages.getMessage('noPromptLong', [], 'demoMode'),
        required: false,
        hidden: true
    })
};
