// salesforce-alm/dist/commands/force/apex/test/run.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const kit_1 = require("@salesforce/kit");
const Messages = require("../../../../lib/messages");
const ReporterCommand_1 = require("../../../../ReporterCommand");
const messages = Messages();
const consts = require("../../../../lib/core/constants");
class ApexTestRunCommand extends ReporterCommand_1.ReporterCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const Command = require('../../../../lib/apex/apexTestCommand');
        const results = await this.execLegacyCommand(new Command(), context);
        if (process.env.SFDX_IMPROVED_CODE_COVERAGE !== 'true' && this.flags.codecoverage) {
            this.ux.warn(messages.getMessage('apexReportImprovedCoverageWarning', [], 'apex'));
        }
        return results;
    }
}
exports.ApexTestRunCommand = ApexTestRunCommand;
ApexTestRunCommand.theDescription = messages.getMessage('apexTestCommandDescription', [], 'apex');
ApexTestRunCommand.longDescription = messages.getMessage('apexTestCommandDescriptionLong', [], 'apex');
ApexTestRunCommand.help = messages.getMessage('apexTestCommandHelp', [], 'apex');
ApexTestRunCommand.requiresProject = false;
ApexTestRunCommand.requiresUsername = true;
ApexTestRunCommand.resultFormatOptions = { options: ['human', 'tap', 'junit', 'json'] };
ApexTestRunCommand.flagsConfig = {
    classnames: command_1.flags.array({
        char: 'n',
        description: messages.getMessage('apexTestCommandParamTestNames', [], 'apex'),
        longDescription: messages.getMessage('apexTestCommandParamTestNamesLong', [], 'apex'),
        required: false,
        exclusive: ['suitenames', 'tests']
    }),
    suitenames: command_1.flags.array({
        char: 's',
        description: messages.getMessage('apexTestCommandParamTestSuites', [], 'apex'),
        longDescription: messages.getMessage('apexTestCommandParamTestSuitesLong', [], 'apex'),
        required: false
    }),
    tests: command_1.flags.array({
        char: 't',
        description: messages.getMessage('apexTestCommandParamTests', [], 'apex'),
        longDescription: messages.getMessage('apexTestCommandParamTestsLong', [], 'apex'),
        required: false
    }),
    codecoverage: command_1.flags.boolean({
        char: 'c',
        description: messages.getMessage('apexCommandParamCodeCoverage', [], 'apex'),
        longDescription: messages.getMessage('apexCommandParamCodeCoverageLong', [], 'apex'),
        required: false
    }),
    outputdir: command_1.flags.directory({
        char: 'd',
        description: messages.getMessage('apexCommandParamTestArtifactDir', [], 'apex'),
        longDescription: messages.getMessage('apexCommandParamTestArtifactDirLong', [], 'apex'),
        required: false
    }),
    testlevel: command_1.flags.enum({
        char: 'l',
        description: messages.getMessage('apexTestCommandParamTestLevel', [], 'apex'),
        longDescription: messages.getMessage('apexTestCommandParamTestLevelLong', [], 'apex'),
        required: false,
        options: ['RunLocalTests', 'RunAllTestsInOrg', 'RunSpecifiedTests']
    }),
    wait: command_1.flags.minutes({
        char: 'w',
        description: messages.getMessage('streamingWait', []),
        longDescription: messages.getMessage('streamingWaitLong', []),
        required: false,
        min: kit_1.Duration.minutes(consts.MIN_STREAM_TIMEOUT_MINUTES)
    }),
    precompilewait: command_1.flags.minutes({
        description: messages.getMessage('precompileDescription', [], 'apexPreCompileCommand'),
        longDescription: messages.getMessage('precompileLongDescription', [], 'apexPreCompileCommand'),
        required: false,
        hidden: true,
        min: kit_1.Duration.minutes(consts.DEFAULT_TIMEOUT.minutes)
    }),
    synchronous: command_1.flags.boolean({
        char: 'y',
        description: messages.getMessage('apexTestCommandParamSynchronous', [], 'apex'),
        longDescription: messages.getMessage('apexTestCommandParamSynchronousLong', [], 'apex'),
        required: false
    }),
    verbose: command_1.flags.builtin({
        description: messages.getMessage('verboseDescription', [], 'apex'),
        longDescription: messages.getMessage('verboseLongDescription', [], 'apex')
    })
};
