// salesforce-alm/dist/commands/force/schema/sobject/describe.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../../lib/force-cli/force-cli-messages");
const ToolbeltCommand_1 = require("../../../../ToolbeltCommand");
class SchemaSobjectDescribeCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const { SchemaSObjectDescribeCommand } = require('../../../../lib/schema/schemaSObjectDescribeCommand');
        return await this.execLegacyCommand(new SchemaSObjectDescribeCommand(), context);
    }
}
exports.SchemaSobjectDescribeCommand = SchemaSobjectDescribeCommand;
SchemaSobjectDescribeCommand.theDescription = Messages.get('SchemaSObjectDescribeDescription');
SchemaSobjectDescribeCommand.longDescription = Messages.get('SchemaSObjectDescribeLongDescription');
SchemaSobjectDescribeCommand.help = Messages.get('SchemaSObjectDescribeHelp');
SchemaSobjectDescribeCommand.requiresProject = false;
SchemaSobjectDescribeCommand.flagsConfig = {
    sobjecttype: command_1.flags.string({
        char: 's',
        required: true,
        hidden: false,
        description: Messages.get('SchemaSObjectDescribeObjectNameDescription'),
        longDescription: Messages.get('SchemaSObjectDescribeObjectNameLongDescription')
    }),
    usetoolingapi: command_1.flags.boolean({
        char: 't',
        required: false,
        hidden: false,
        description: Messages.get('SchemaSObjectDescribeToolingDescription'),
        longDescription: Messages.get('SchemaSObjectDescribeToolingLongDescription')
    })
};
SchemaSobjectDescribeCommand.requiresUsername = true;
