// salesforce-alm/dist/commands/force/schema/sobject/list.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../../lib/force-cli/force-cli-messages");
const ToolbeltCommand_1 = require("../../../../ToolbeltCommand");
class SchemaSobjectListCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const { SchemaSObjectListCommand } = require('../../../../lib/schema/schemaSObjectListCommand');
        return await this.execLegacyCommand(new SchemaSObjectListCommand(), context);
    }
}
exports.SchemaSobjectListCommand = SchemaSobjectListCommand;
SchemaSobjectListCommand.theDescription = Messages.get('SchemaSObjectListDescription');
SchemaSobjectListCommand.longDescription = Messages.get('SchemaSObjectListLongDescription');
SchemaSobjectListCommand.help = Messages.get('SchemaSObjectListHelp');
SchemaSobjectListCommand.requiresProject = false;
SchemaSobjectListCommand.flagsConfig = {
    sobjecttypecategory: command_1.flags.string({
        char: 'c',
        required: true,
        hidden: false,
        description: Messages.get('SchemaSObjectListTypeDescription'),
        longDescription: Messages.get('SchemaSObjectListTypeLongDescription')
    })
};
SchemaSobjectListCommand.requiresUsername = true;
