// @lwc/synthetic-shadow/lib/faux-shadow/traverse.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const assert_1 = __importDefault(require("../shared/assert"));
const node_1 = require("./node");
const node_2 = require("../env/node");
const element_1 = require("../env/element");
const language_1 = require("../shared/language");
const language_2 = require("../shared/language");
const shadow_root_1 = require("./shadow-root");
const slot_1 = require("./slot");
require("../polyfills/node-get-root-node/main");
// Extract the patched getRootNode
exports.patchedGetRootNode = Node.prototype.getRootNode;
function getNodeOwner(node) {
    if (!(node instanceof Node)) {
        return null;
    }
    const ownerKey = node_1.getNodeNearestOwnerKey(node);
    if (language_1.isUndefined(ownerKey)) {
        return null;
    }
    let nodeOwner = node;
    // At this point, node is a valid node with owner identity, now we need to find the owner node
    // search for a custom element with a VM that owns the first element with owner identity attached to it
    while (!language_2.isNull(nodeOwner) && node_1.getNodeKey(nodeOwner) !== ownerKey) {
        nodeOwner = node_2.parentNodeGetter.call(nodeOwner);
    }
    if (language_2.isNull(nodeOwner)) {
        return null;
    }
    return nodeOwner;
}
exports.getNodeOwner = getNodeOwner;
function isSlotElement(elm) {
    return element_1.tagNameGetter.call(elm) === 'SLOT';
}
exports.isSlotElement = isSlotElement;
function isNodeOwnedBy(owner, node) {
    if (process.env.NODE_ENV !== 'production') {
        assert_1.default.invariant(owner instanceof HTMLElement, `isNodeOwnedBy() should be called with an element as the first argument instead of ${owner}`);
        assert_1.default.invariant(node instanceof Node, `isNodeOwnedBy() should be called with a node as the second argument instead of ${node}`);
        assert_1.default.isTrue(node_2.compareDocumentPosition.call(node, owner) & node_2.DOCUMENT_POSITION_CONTAINS, `isNodeOwnedBy() should never be called with a node that is not a child node of ${owner}`);
    }
    const ownerKey = node_1.getNodeNearestOwnerKey(node);
    return language_1.isUndefined(ownerKey) || node_1.getNodeKey(owner) === ownerKey;
}
exports.isNodeOwnedBy = isNodeOwnedBy;
function shadowRootChildNodes(root) {
    const elm = shadow_root_1.getHost(root);
    return getAllMatches(elm, node_2.childNodesGetter.call(elm));
}
exports.shadowRootChildNodes = shadowRootChildNodes;
function getAllMatches(owner, nodeList) {
    const filteredAndPatched = [];
    for (let i = 0, len = nodeList.length; i < len; i += 1) {
        const node = nodeList[i];
        const isOwned = isNodeOwnedBy(owner, node);
        if (isOwned) {
            // Patch querySelector, querySelectorAll, etc
            // if element is owned by VM
            language_1.ArrayPush.call(filteredAndPatched, node);
        }
    }
    return filteredAndPatched;
}
exports.getAllMatches = getAllMatches;
function getFirstMatch(owner, nodeList) {
    for (let i = 0, len = nodeList.length; i < len; i += 1) {
        if (isNodeOwnedBy(owner, nodeList[i])) {
            return nodeList[i];
        }
    }
    return null;
}
exports.getFirstMatch = getFirstMatch;
function shadowRootQuerySelector(root, selector) {
    const elm = shadow_root_1.getHost(root);
    const nodeList = element_1.querySelectorAll.call(elm, selector);
    return getFirstMatch(elm, nodeList);
}
exports.shadowRootQuerySelector = shadowRootQuerySelector;
function shadowRootQuerySelectorAll(root, selector) {
    const elm = shadow_root_1.getHost(root);
    const nodeList = element_1.querySelectorAll.call(elm, selector);
    return getAllMatches(elm, nodeList);
}
exports.shadowRootQuerySelectorAll = shadowRootQuerySelectorAll;
function getFilteredChildNodes(node) {
    let children;
    if (!language_1.isUndefined(node_1.getNodeKey(node))) {
        // node itself is a custom element
        // lwc element, in which case we need to get only the nodes
        // that were slotted
        const slots = element_1.querySelectorAll.call(node, 'slot');
        children = language_1.ArrayReduce.call(slots, (seed, slot) => {
            if (isNodeOwnedBy(node, slot)) {
                language_1.ArrayPush.apply(seed, slot_1.getFilteredSlotAssignedNodes(slot));
            }
            return seed;
        }, []);
    }
    else {
        // regular element
        children = node_2.childNodesGetter.call(node);
    }
    const owner = getNodeOwner(node);
    if (language_2.isNull(owner)) {
        return [];
    }
    // Typescript is inferring the wrong function type for this particular
    // overloaded method: https://github.com/Microsoft/TypeScript/issues/27972
    // @ts-ignore type-mismatch
    return language_1.ArrayReduce.call(children, (seed, child) => {
        if (isNodeOwnedBy(owner, child)) {
            language_1.ArrayPush.call(seed, child);
        }
        return seed;
    }, []);
}
exports.getFilteredChildNodes = getFilteredChildNodes;
