// salesforce-alm/dist/commands/force/source/convert.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../lib/messages");
const ToolbeltCommand_1 = require("../../../ToolbeltCommand");
const messages = Messages();
class SourceConvertCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const LegacySourceConvertCommand = require('../../../lib/source/sourceConvertCommand');
        return this.execLegacyCommand(new LegacySourceConvertCommand(), context);
    }
}
exports.SourceConvertCommand = SourceConvertCommand;
SourceConvertCommand.theDescription = messages.getMessage('description', null, 'source_convert');
SourceConvertCommand.longDescription = messages.getMessage('longDescription', null, 'source_convert');
SourceConvertCommand.help = messages.getMessage('help', null, 'source_convert');
SourceConvertCommand.initializeMetadataRegistry = true;
SourceConvertCommand.requiresUsername = false;
SourceConvertCommand.requiresProject = true;
SourceConvertCommand.showProgress = true;
SourceConvertCommand.flagsConfig = {
    rootdir: command_1.flags.directory({
        char: 'r',
        description: messages.getMessage('rootParam', null, 'source_convert'),
        longDescription: messages.getMessage('rootParamLongDescription', null, 'source_convert'),
        required: false
    }),
    outputdir: command_1.flags.directory({
        char: 'd',
        description: messages.getMessage('outputDirectoryParam', null, 'source_convert'),
        longDescription: messages.getMessage('outputDirectoryParamLongDescription', null, 'source_convert'),
        required: false
    }),
    packagename: command_1.flags.string({
        char: 'n',
        description: messages.getMessage('packageNameParam', null, 'source_convert'),
        longDescription: messages.getMessage('packageNameParamLongDescription', null, 'source_convert'),
        required: false
    }),
    manifest: command_1.flags.string({
        char: 'x',
        description: messages.getMessage('manifestDescription', null, 'source_convert'),
        longDescription: messages.getMessage('manifestLongDescription', null, 'source_convert'),
        required: false
    }),
    sourcepath: command_1.flags.array({
        char: 'p',
        required: false,
        hidden: false,
        description: messages.getMessage('sourcePathDescription', null, 'source_convert'),
        longDescription: messages.getMessage('sourcePathLongDescription', null, 'source_convert'),
        exclusive: ['manifest', 'metadata']
    }),
    metadata: command_1.flags.array({
        char: 'm',
        description: messages.getMessage('metadataParamDescription', null, 'source_convert'),
        longDescription: messages.getMessage('metadataParamLongDescription', null, 'source_convert'),
        required: false,
        exclusive: ['manifest', 'sourcepath']
    })
};
