// salesforce-alm/dist/commands/force/source/status.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../lib/messages");
const ToolbeltCommand_1 = require("../../../ToolbeltCommand");
const messages = Messages();
class SourceStatusCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const _ = require('lodash');
        const context = await this.resolveLegacyContext();
        const SrcStatusCommand = require('../../../lib/source/srcStatusCommand');
        if ((_.isNil(context.flags.local) && _.isNil(context.flags.remote)) || context.flags.all) {
            context.local = true;
            context.remote = true;
        }
        else if (context.flags.local) {
            context.local = true;
        }
        else if (context.flags.remote) {
            context.remote = true;
        }
        const command = new SrcStatusCommand(context);
        return this.execLegacyCommand(command, context);
    }
}
exports.SourceStatusCommand = SourceStatusCommand;
SourceStatusCommand.theDescription = messages.getMessage('statusCommandCliDescription');
SourceStatusCommand.longDescription = messages.getMessage('statusCommandCliLongDescription');
SourceStatusCommand.help = messages.getMessage('statusCommandCliHelp');
SourceStatusCommand.showProgress = true;
SourceStatusCommand.requiresProject = true;
SourceStatusCommand.requiresUsername = true;
SourceStatusCommand.flagsConfig = {
    all: command_1.flags.boolean({
        char: 'a',
        description: messages.getMessage('statusCommandAllOptionDescription'),
        longDescription: messages.getMessage('statusCommandAllOptionDescriptionLong'),
        required: false
    }),
    local: command_1.flags.boolean({
        char: 'l',
        description: messages.getMessage('statusCommandLocalOptionDescription'),
        longDescription: messages.getMessage('statusCommandLocalOptionDescriptionLong'),
        required: false
    }),
    remote: command_1.flags.boolean({
        char: 'r',
        description: messages.getMessage('statusCommandRemoteOptionDescription'),
        longDescription: messages.getMessage('statusCommandRemoteOptionDescriptionLong'),
        required: false
    })
};
