// @lwc/engine/lib/env/node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const language_1 = require("../shared/language");
const { appendChild, insertBefore, removeChild, replaceChild } = Node.prototype;
exports.appendChild = appendChild;
exports.insertBefore = insertBefore;
exports.removeChild = removeChild;
exports.replaceChild = replaceChild;
const parentNodeGetter = language_1.getOwnPropertyDescriptor(Node.prototype, 'parentNode').get;
exports.parentNodeGetter = parentNodeGetter;
const parentElementGetter = language_1.hasOwnProperty.call(Node.prototype, 'parentElement')
    ? language_1.getOwnPropertyDescriptor(Node.prototype, 'parentElement').get
    : language_1.getOwnPropertyDescriptor(HTMLElement.prototype, 'parentElement').get; // IE11
exports.parentElementGetter = parentElementGetter;
