// salesforce-alm/dist/commands/force/apex/execute.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../lib/force-cli/force-cli-messages");
const ToolbeltCommand_1 = require("../../../ToolbeltCommand");
class ApexExecuteCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const { ApexExecuteCommand } = require('../../../lib/apex/apexExecuteCommand');
        return await this.execLegacyCommand(new ApexExecuteCommand(), context);
    }
}
exports.ApexExecuteCommand = ApexExecuteCommand;
ApexExecuteCommand.theDescription = Messages.get('ApexExecDescription');
ApexExecuteCommand.longDescription = Messages.get('ApexExecLongDescription');
ApexExecuteCommand.help = Messages.get('ApexExecHelp');
ApexExecuteCommand.requiresProject = false;
ApexExecuteCommand.flagsConfig = {
    apexcodefile: command_1.flags.filepath({
        char: 'f',
        required: false,
        hidden: false,
        description: Messages.get('ApexExecFilePathDescription'),
        longDescription: Messages.get('ApexExecFilePathLongDescription')
    })
};
ApexExecuteCommand.requiresUsername = true;
