// salesforce-alm/dist/commands/force/source/deploy/report.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const ToolbeltCommand_1 = require("../../../../ToolbeltCommand");
const core_1 = require("@salesforce/core");
const command_1 = require("@salesforce/command");
const kit_1 = require("@salesforce/kit");
const consts = require("../../../../lib/core/constants");
const Stash = require("../../../../lib/core/stash");
const { DEFAULT_SRC_WAIT_MINUTES, MIN_SRC_WAIT_MINUTES } = consts;
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('salesforce-alm', 'source_deploy_report');
const commonMsgs = core_1.Messages.loadMessages('salesforce-alm', 'source');
class SourceDeployReportCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const MdapiDeployReportCommand = require('../../../../lib/mdapi/mdapiDeployReportCommand');
        return this.execLegacyCommand(new MdapiDeployReportCommand(context, Stash.Commands.SOURCE_DEPLOY), context);
    }
}
exports.SourceDeployReportCommand = SourceDeployReportCommand;
SourceDeployReportCommand.theDescription = messages.getMessage('description');
SourceDeployReportCommand.longDescription = messages.getMessage('longDescription');
SourceDeployReportCommand.help = messages.getMessage('help');
SourceDeployReportCommand.requiresProject = false;
SourceDeployReportCommand.requiresUsername = true;
SourceDeployReportCommand.flagsConfig = {
    wait: command_1.flags.minutes({
        char: 'w',
        required: false,
        hidden: false,
        default: kit_1.Duration.minutes(DEFAULT_SRC_WAIT_MINUTES),
        min: kit_1.Duration.minutes(MIN_SRC_WAIT_MINUTES),
        description: commonMsgs.getMessage('waitParamDescription', [DEFAULT_SRC_WAIT_MINUTES]),
        longDescription: commonMsgs.getMessage('waitParamDescriptionLong')
    }),
    jobid: command_1.flags.id({
        char: 'i',
        description: messages.getMessage('jobId'),
        longDescription: messages.getMessage('jobIdLong'),
        required: false
    })
};
