// salesforce-alm/dist/lib/source/decompositionStrategy/customObjectTranslationDecompositionConfig.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Decomposition configuration for the describeMetadata decomposition and the folder per subtype
 * workspace manifestation.
 */
class CustomObjectTranslationDecompositionConfig {
    constructor(metadataName, isGlobal, isEmptyContainer) {
        this.strategy = 'describeMetadata';
        this.workspaceStrategy = 'folderPerSubtype';
        this.commitStrategy = 'virtualDecomposition';
        this.contentStrategy = 'N/A';
        this.metadataName = metadataName;
        this.isGlobal = isGlobal;
        this.isEmptyContainer = isEmptyContainer;
        this.useSparseComposition = false;
        this.decompositions = [];
    }
}
exports.CustomObjectTranslationDecompositionConfig = CustomObjectTranslationDecompositionConfig;
