// salesforce-alm/dist/lib/source/decompositionStrategy/staticResourceContentStrategy.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const staticResource_1 = require("./staticResource");
class StaticResourceContentStrategy {
    constructor(metadataType, metadataRegistry, workspaceVersion) {
        this.metadataType = metadataType;
        this.metadataRegistry = metadataRegistry;
        this.workspaceVersion = workspaceVersion;
    }
    getContentPaths(metadataFilePath) {
        const staticResource = new staticResource_1.StaticResource(metadataFilePath, this.metadataType, this.workspaceVersion);
        return staticResource.getContentPaths();
    }
    saveContent(metadataFilePath, retrievedContentFilePaths, retrievedMetadataFilePath, createDuplicates, unsupportedMimeTypes, forceoverwrite = false) {
        const newPaths = [];
        const deletedPaths = [];
        const staticResource = new staticResource_1.StaticResource(metadataFilePath, this.metadataType, this.workspaceVersion, retrievedMetadataFilePath, unsupportedMimeTypes);
        const [updatedPaths, duplicatePaths] = staticResource.saveResource(retrievedContentFilePaths[0], createDuplicates, forceoverwrite);
        return [newPaths, updatedPaths, deletedPaths, duplicatePaths];
    }
}
exports.StaticResourceContentStrategy = StaticResourceContentStrategy;
