// @lwc/compiler/dist/commonjs/rollup-plugins/module-resolver.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const path = __importStar(require("path"));
const errors_1 = require("@lwc/errors");
const EMPTY_IMPLICIT_CSS_CONTENT = '';
const EMPTY_IMPLICIT_HTML_CONTENT = 'export default void 0';
const IMPLICIT_DEFAULT_HTML_PATH = '@lwc/resources/empty_html.js';
const IMPLICIT_DEFAULT_CSS_PATH = '@lwc/resources/empty_css.css';
const VALID_EXTENSIONS = ['.js', '.ts', '.html', '.css']; // order of resolution is important
function isRelativeImport(id) {
    return id.startsWith('.');
}
function isImplicitCssImport(id, importee) {
    return (path.extname(id) === '.css' &&
        path.extname(importee) === '.html' &&
        path.basename(id, '.css') === path.basename(importee, '.html'));
}
function isImplicitHTMLImport(importee, importer) {
    return (importer &&
        path.extname(importer) === '.js' &&
        path.extname(importee) === '.html' &&
        path.dirname(importer) === path.dirname(importee) &&
        path.basename(importer, '.js') === path.basename(importee, '.html'));
}
function isFirstCharacterUppercased(importee) {
    const upperCaseRegex = /^[A-Z]/;
    return importee && upperCaseRegex.test(importee);
}
function inferExtension(fileName, files) {
    if (!path.extname(fileName)) {
        const ext = VALID_EXTENSIONS.find(ext => files.hasOwnProperty(fileName + ext)) || '';
        return fileName + ext;
    }
    return fileName;
}
function fileExists(fileName, { files }) {
    return files.hasOwnProperty(fileName);
}
function readFile(filename, options) {
    const { files } = options;
    if (fileExists(filename, options)) {
        return files[filename];
    }
    else {
        throw errors_1.generateCompilerError(errors_1.ModuleResolutionErrors.NONEXISTENT_FILE, {
            messageArgs: [filename],
            origin: { filename },
        });
    }
}
function generateModuleResolutionError(importee, importer, options) {
    const absPath = getAbsolutePath(importee, importer, options);
    const caseIgnoredFilename = getCaseIgnoredFilenameMatch(options.files, absPath);
    return caseIgnoredFilename
        ? errors_1.generateCompilerError(errors_1.ModuleResolutionErrors.IMPORT_AND_FILE_NAME_CASE_MISMATCH, {
            messageArgs: [
                importee,
                importer,
                caseIgnoredFilename.substr(0, caseIgnoredFilename.length - path.extname(caseIgnoredFilename).length),
            ],
            origin: { filename: importer },
        })
        : errors_1.generateCompilerError(errors_1.ModuleResolutionErrors.IMPORTEE_RESOLUTION_FROM_IMPORTER_FAILED, {
            messageArgs: [importee, importer, absPath],
            origin: { filename: importer },
        });
}
function generateEntryResolutionError(importee, importer, options) {
    const absPath = getAbsolutePath(importee, importer, options);
    const caseIgnoredFilename = getCaseIgnoredFilenameMatch(options.files, absPath);
    return caseIgnoredFilename
        ? errors_1.generateCompilerError(errors_1.ModuleResolutionErrors.FOLDER_AND_FILE_NAME_CASE_MISMATCH, {
            messageArgs: [caseIgnoredFilename, importee],
            origin: { filename: importer },
        })
        : errors_1.generateCompilerError(errors_1.ModuleResolutionErrors.IMPORTEE_RESOLUTION_FAILED, {
            messageArgs: [importee],
            origin: { filename: importer },
        });
}
function getAbsolutePath(importee, importer, options) {
    const { baseDir, files } = options;
    const relPath = importer ? path.dirname(importer) : baseDir || '';
    return inferExtension(path.join(relPath, importee), files);
}
function getCaseIgnoredFilenameMatch(files, nameToMatch) {
    return Object.keys(files).find((bundleFile) => bundleFile.toLowerCase() === nameToMatch);
}
function default_1({ options }) {
    return {
        name: 'lwc-module-resolver',
        resolveId(importee, importer) {
            // Mark non-relative imports (eg. 'lwc' or 'x/foo') as external dependencies.
            if (!isRelativeImport(importee) && importer) {
                return false;
            }
            if (isFirstCharacterUppercased(importee)) {
                throw errors_1.generateCompilerError(errors_1.ModuleResolutionErrors.FOLDER_NAME_STARTS_WITH_CAPITAL_LETTER, {
                    messageArgs: [
                        importee,
                        importee.charAt(0).toLowerCase() + importee.slice(1),
                    ],
                });
            }
            const absPath = getAbsolutePath(importee, importer, options);
            if (!fileExists(absPath, options)) {
                if (isImplicitCssImport(importee, importer)) {
                    return IMPLICIT_DEFAULT_CSS_PATH;
                }
                if (isImplicitHTMLImport(absPath, importer)) {
                    return IMPLICIT_DEFAULT_HTML_PATH;
                }
                throw importer
                    ? generateModuleResolutionError(importee, importer, options)
                    : generateEntryResolutionError(importee, importer, options);
            }
            return absPath;
        },
        load(id) {
            if (id === IMPLICIT_DEFAULT_CSS_PATH) {
                return EMPTY_IMPLICIT_CSS_CONTENT;
            }
            if (id === IMPLICIT_DEFAULT_HTML_PATH) {
                return EMPTY_IMPLICIT_HTML_CONTENT;
            }
            return path.extname(id) === '.css' && !fileExists(id, options)
                ? EMPTY_IMPLICIT_CSS_CONTENT
                : readFile(id, options);
        },
    };
}
exports.default = default_1;
