// salesforce-alm/dist/commands/force/package/uninstall.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../lib/messages");
const ToolbeltCommand_1 = require("../../../ToolbeltCommand");
const messages = Messages();
const consts = require("../../../lib/core/constants");
class PackageUninstallCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const PackageUninstallCommand = require('../../../lib/package/packageUninstallCommand');
        return this.execLegacyCommand(new PackageUninstallCommand(), context);
    }
}
exports.PackageUninstallCommand = PackageUninstallCommand;
PackageUninstallCommand.theDescription = messages.getMessage('cliDescription', [], 'package_uninstall');
PackageUninstallCommand.longDescription = messages.getMessage('cliDescriptionLong', [], 'package_uninstall');
PackageUninstallCommand.help = messages.getMessage('help', [], 'package_uninstall');
PackageUninstallCommand.requiresProject = false;
PackageUninstallCommand.orgType = consts.DEFAULT_USERNAME;
PackageUninstallCommand.requiresUsername = true;
PackageUninstallCommand.flagsConfig = {
    wait: command_1.flags.minutes({
        char: 'w',
        description: messages.getMessage('wait', [], 'package_uninstall'),
        longDescription: messages.getMessage('waitLong', [], 'package_uninstall'),
        required: false
    }),
    package: command_1.flags.string({
        char: 'p',
        description: messages.getMessage('package', [], 'package_uninstall'),
        longDescription: messages.getMessage('packageLong', [], 'package_uninstall'),
        required: false
    })
};
