// @salesforce/plugin-analytics/lib/uploader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@salesforce/core");
const telemetry_1 = require("@salesforce/telemetry");
const ts_types_1 = require("@salesforce/ts-types");
const analytics_1 = require("./analytics");
const debuger_1 = require("./debuger");
const PROJECT = 'salesforce-cli';
const APP_INSIGHTS_KEY = '2ca64abb-6123-4c7b-bd9e-4fe73e71fe9c';
class Uploader {
    constructor(analytics) {
        this.analytics = analytics;
    }
    /**
     * Sends events from analytics to telemetry.
     */
    static async upload(cacheDir, analyticsFilePath) {
        const analytics = (global.cliTelemetry = await analytics_1.default.create({ cacheDir, analyticsFilePath }));
        const uploader = new Uploader(analytics);
        await uploader.sendToTelemetry();
    }
    /**
     * Reads the analytics events from file and sends them to the telemetry service.
     */
    async sendToTelemetry() {
        let reporter;
        try {
            reporter = await telemetry_1.default.create({
                project: PROJECT,
                key: APP_INSIGHTS_KEY,
                userId: this.analytics.getCLIId(),
                waitForConnection: true
            });
        }
        catch (error) {
            debuger_1.debug(`Error creating reporter: ${error.message}`);
            // We can't do much without a reporter, so clear the analytics file and move on.
            await this.analytics.clear();
            return;
        }
        try {
            const events = await this.analytics.read();
            for (const event of events) {
                const eventType = ts_types_1.asString(event.type) || analytics_1.default.EVENT;
                const eventName = ts_types_1.asString(event.eventName) || 'UNKNOWN';
                delete event.type;
                delete event.eventName;
                if (eventType === analytics_1.default.EVENT) {
                    // Resolve orgs for all events.
                    event.orgId = await this.getOrgId(false, ts_types_1.asString(event.orgUsername));
                    event.devHubId = await this.getOrgId(true, ts_types_1.asString(event.devHubUsername));
                    // Usernames are GDPR info, so don't log.
                    delete event.orgUsername;
                    delete event.devHubUsername;
                    reporter.sendTelemetryEvent(eventName, event);
                }
                else if (eventType === analytics_1.default.EXCEPTION) {
                    const error = new Error();
                    // We know this is an object because it is logged as such
                    const errorObject = event.error;
                    delete event.error;
                    Object.assign(error, errorObject);
                    error.name = ts_types_1.asString(errorObject.name) || 'Unknown';
                    error.message = ts_types_1.asString(errorObject.messaage) || 'Unknown';
                    error.stack = ts_types_1.asString(errorObject.stack) || 'Unknown';
                    reporter.sendTelemetryException(error, event);
                }
            }
        }
        catch (error) {
            debuger_1.debug(`Error reading or sending telemetry events: ${error.message}`);
        }
        finally {
            try {
                // We are done sending events
                reporter.stop();
            }
            catch (error) {
                debuger_1.debug(`Error stopping telemetry reporter: ${error.message}`);
            }
            finally {
                // We alway want to clear the file.
                await this.analytics.clear();
            }
        }
    }
    async getOrgId(isDevHub, aliasOrUsername) {
        const orgOptions = { isDevHub };
        // If no aliasOrUsername, it will try to get the default
        if (aliasOrUsername) {
            orgOptions.aliasOrUsername = aliasOrUsername;
        }
        try {
            return (await core_1.Org.create(orgOptions)).getOrgId();
        }
        catch (error) {
            // Don't do anything, there is no specified or default.
        }
    }
}
exports.Uploader = Uploader;
