// salesforce-alm/dist/commands/force/config/set.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../lib/messages");
const ToolbeltCommand_1 = require("../../../ToolbeltCommand");
const messages = Messages();
class ConfigSetCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const ConfigSetCommand = require('../../../lib/config/ConfigSetCommand');
        return this.execLegacyCommand(new ConfigSetCommand(), context);
    }
}
exports.ConfigSetCommand = ConfigSetCommand;
ConfigSetCommand.theDescription = messages.getMessage('description', [], 'configSetCommand');
ConfigSetCommand.longDescription = messages.getMessage('descriptionLong', [], 'configSetCommand');
ConfigSetCommand.help = messages.getMessage('help', [], 'configSetCommand');
ConfigSetCommand.requiresProject = false;
ConfigSetCommand.varargs = { required: true };
ConfigSetCommand.flagsConfig = {
    global: command_1.flags.boolean({
        char: 'g',
        description: messages.getMessage('global', [], 'configSetCommand'),
        longDescription: messages.getMessage('globalLong', [], 'configSetCommand'),
        required: false
    })
};
