// salesforce-alm/dist/commands/force/user/create.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const messages = require("../../../lib/messages");
const ToolbeltCommand_1 = require("../../../ToolbeltCommand");
const { getMessage } = messages();
class UserCreateCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const { UserCreateCommand } = require('../../../lib/user/userCreateCommand');
        return await this.execLegacyCommand(new UserCreateCommand(), context);
    }
}
exports.UserCreateCommand = UserCreateCommand;
UserCreateCommand.theDescription = getMessage('description', [], 'user_create');
UserCreateCommand.longDescription = getMessage('longDescription', [], 'user_create');
UserCreateCommand.help = getMessage('help', [], 'user_create');
UserCreateCommand.requiresProject = false;
UserCreateCommand.varargs = true;
UserCreateCommand.requiresUsername = true;
UserCreateCommand.supportsDevhubUsername = true;
UserCreateCommand.flagsConfig = {
    definitionfile: command_1.flags.filepath({
        char: 'f',
        description: getMessage('definitionfileParam', [], 'user_create'),
        longDescription: getMessage('definitionfileParamLong', [], 'user_create'),
        required: false
    }),
    setalias: command_1.flags.string({
        char: 'a',
        description: getMessage('setaliasParam', [], 'user_create'),
        longDescription: getMessage('setaliasParamLong', [], 'user_create'),
        required: false
    })
};
