// salesforce-alm/dist/commands/force/mdapi/deploy/cancel.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const kit_1 = require("@salesforce/kit");
const ToolbeltCommand_1 = require("../../../../ToolbeltCommand");
const consts = require("../../../../lib/core/constants");
const { DEFAULT_SRC_WAIT_MINUTES, MIN_SRC_WAIT_MINUTES } = consts;
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('salesforce-alm', 'mdapi_cancel');
const commonMsgs = core_1.Messages.loadMessages('salesforce-alm', 'source');
class MdapiDeployCancelCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const { MdapiCancelCommand } = require('../../../../lib/mdapi/mdapiDeployCancelCommand');
        const context = await this.resolveLegacyContext();
        return await this.execLegacyCommand(new MdapiCancelCommand(), context);
    }
}
exports.MdapiDeployCancelCommand = MdapiDeployCancelCommand;
MdapiDeployCancelCommand.theDescription = messages.getMessage('description');
MdapiDeployCancelCommand.longDescription = messages.getMessage('longDescription');
MdapiDeployCancelCommand.help = messages.getMessage('help');
MdapiDeployCancelCommand.requiresProject = false;
MdapiDeployCancelCommand.requiresUsername = true;
MdapiDeployCancelCommand.flagsConfig = {
    wait: command_1.flags.minutes({
        char: 'w',
        required: false,
        hidden: false,
        default: kit_1.Duration.minutes(DEFAULT_SRC_WAIT_MINUTES),
        min: kit_1.Duration.minutes(MIN_SRC_WAIT_MINUTES),
        description: commonMsgs.getMessage('waitParamDescription', [DEFAULT_SRC_WAIT_MINUTES]),
        longDescription: commonMsgs.getMessage('waitParamDescriptionLong')
    }),
    jobid: command_1.flags.id({
        char: 'i',
        description: messages.getMessage('jobId'),
        longDescription: messages.getMessage('jobIdLong'),
        required: false
    })
};
