// @lwc/template-compiler/dist/commonjs/shared/ir.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isComponentProp = exports.isSlot = exports.isTemplate = exports.isCustomElement = exports.isCommentNode = exports.isTextNode = exports.isElement = exports.createComment = exports.createText = exports.createElement = void 0;
function createElement(original, parent) {
    return {
        type: 'element',
        tag: original.tagName,
        namespace: original.namespaceURI,
        attrsList: original.attrs,
        parent,
        children: [],
        __original: original,
    };
}
exports.createElement = createElement;
function createText(original, parent, value) {
    return {
        type: 'text',
        parent,
        value,
        __original: original,
    };
}
exports.createText = createText;
function createComment(original, parent, value) {
    return {
        type: 'comment',
        parent,
        value,
        __original: original,
    };
}
exports.createComment = createComment;
function isElement(node) {
    return node.type === 'element';
}
exports.isElement = isElement;
function isTextNode(node) {
    return node.type === 'text';
}
exports.isTextNode = isTextNode;
function isCommentNode(node) {
    return node.type === 'comment';
}
exports.isCommentNode = isCommentNode;
function isCustomElement(node) {
    return isElement(node) && node.component !== undefined;
}
exports.isCustomElement = isCustomElement;
function isTemplate(element) {
    return element.tag === 'template';
}
exports.isTemplate = isTemplate;
function isSlot(element) {
    return element.tag === 'slot';
}
exports.isSlot = isSlot;
function isComponentProp(identifier, root) {
    var _a;
    const { name } = identifier;
    let current = root;
    // Walking up the AST and checking for each node to find if the identifer name is identical to
    // an iteration variable.
    while (current !== undefined) {
        if (isElement(current)) {
            const { forEach, forOf } = current;
            if ((forEach === null || forEach === void 0 ? void 0 : forEach.item.name) === name ||
                ((_a = forEach === null || forEach === void 0 ? void 0 : forEach.index) === null || _a === void 0 ? void 0 : _a.name) === name ||
                (forOf === null || forOf === void 0 ? void 0 : forOf.iterator.name) === name) {
                return false;
            }
        }
        current = current.parent;
    }
    // The identifier is bound to a component property if no match is found after reaching to AST
    // root.
    return true;
}
exports.isComponentProp = isComponentProp;
